stac/harness/ilp/template.zpl

```diff
@@ -0,0 +1,77 @@
+
+# Template with relation handling
+
+# Automatically generated
+# param EDU_COUNT := ??? ;
+# param TURN_COUNT := ??? ;
+# param LABEL_COUNT := ??? ;
+
+set EDUs := {1 to EDU_COUNT} ;
+set Turns := {1 to TURN_COUNT} ;
+set Labels := {1 to LABEL_COUNT} ;
+
+param TLEN[Turns] := read "./turn.dat" as "n+" ;
+param TOFF[Turns] := read "./turn.dat" as "n+" skip 1 ;
+param TEDU[EDUs] := read "./turn.dat" as "n+" skip 2 ;
+set TIND[<t> in Turns] := {1 to TLEN[t]} ;
+
+param PATT[EDUs*EDUs] := read "./raw.attach.dat" as "n+" ;
+param PLAB[EDUs*EDUs*Labels] := read "./raw.label.dat" as "n+" ;
+
+var c[<t,i> in Turns*EDUs
+    with i <= TLEN[t]] integer <= EDU_COUNT;
+var h[<i> in EDUs] binary ;
+var a[<i,j> in EDUs*EDUs] binary ;
+var x[<i,j,r> in EDUs*EDUs*Labels] binary ;
+
+# Objective function
+maximize score: sum <i,j,r> in EDUs*EDUs*Labels: PLAB[i, j, r]*x[i, j, r] ;
+
+# Attachment definition
+subto attachment:
+    forall <i,j> in EDUs*EDUs:
+        a[i, j] == sum <r> in Labels: x[i, j, r] ;
+
+# No auto-link
+subto no_diagonal:
+    forall <i> in EDUs:
+        a[i, i] == 0 ;
+
+# No zero-prob links
+subto no_zero_att:
+    forall <i,j> in EDUs*EDUs:
+        if PATT[i, j] == 0
```

Comment:
There is an implicit assumption here that scores are probabilities.
If this is the case, you should write this explicitly in `ilp.py` and add some checks where appropriate to prevent calling the ILP decoder with a non-probabilistic classifier.

Also, this line is the only use of the attachment scores in the ILP solver.
So, it seems to me that the optimization is on the labelling scores only.
If I am right, this means your decoder does not really use the attachment classifier, which is really different from what we have been doing until now, and this approach could be ported trivially to the other decoders.

In any case, what you are doing here is not clear and needs to be thoroughly documented.